src/scripting_api/util.js

```diff
@@ -20,11 +20,753 @@ class Util extends PDFObject {
+      },
+      h: data => {
+        return (1 + ((data.hours + 11) % 12)).toString();
+      },
+      kk: data => {
+        return (data.hours % 12).toString().padStart(2, "0");
+      },
+      k: data => {
+        return (data.hours % 12).toString();
+      },
+      KK: data => {
+        return (1 + ((data.hours + 23) % 24)).toString().padStart(2, "0");
+      },
+      K: data => {
+        return (1 + ((data.hours + 23) % 24)).toString();
+      },
+      mm: data => {
+        return data.minutes.toString().padStart(2, "0");
+      },
+      m: data => {
+        return data.minutes.toString();
+      },
+      SS: data => {
+        return data.seconds.toString().padStart(2, "0");
+      },
+      S: data => {
+        return data.seconds.toString();
+      },
+      FFF: data => {
+        return data.milliseconds.toString().padStart(3, "0");
+      },
+      A: data => {
+        return data.hours < 12 ? "AM" : "PM";
+      },
+      Z: data => {
+        return data.cDate
+          .toLocaleDateString(undefined, { timeZoneName: "short" })
+          .split(" ")[1];
+      },
+      zz: data => {
+        return this._getTZOffset(data.cDate, ":");
+      },
+      z: data => {
+        return this._getTZOffset(data.cDate, "");
+      },
+    };
+
+    const year = cDate.getFullYear();
+    const month = cDate.getMonth();
+    const day = cDate.getDate();
+    const dayOfWeek = cDate.getDay();
+    const hours = cDate.getHours();
+    const minutes = cDate.getMinutes();
+    const seconds = cDate.getSeconds();
+    const milliseconds = cDate.getMilliseconds();
+    const data = {
+      year,
+      month,
+      day,
+      dayOfWeek,
+      hours,
+      minutes,
+      seconds,
+      milliseconds,
+      cDate,
+    };
+
+    const patterns = /(MMMM|MMM|MM|M|DD|D|YYYY|YY|JJJ|J|EEEE|EEE|E|e|G|w|WW|hh|h|kk|k|HH|H|KK|K|mm|m|SS|S|FFF|A|Z|zz|z|\\.)/g;
+    return cFormat.replace(patterns, function (match, pattern) {
+      if (pattern in handlers) {
+        return handlers[pattern](data);
+      }
+      return pattern.charCodeAt(1);
+    });
+  }
+
+  printx(cFormat, cSource) {
+    // case
+    const handlers = [x => x, x => x.toUpperCase(), x => x.toLowerCase()];
+
+    // limits
+    const [LA, LZ, UA, UZ, ZER, NIN] = Array.from("azAZ09").map(c =>
+      c.charCodeAt(0)
+    );
+
+    const buf = [];
+    let i = 0;
+    const ii = cSource.length;
+    let currCase = handlers[0];
+    let escaped = false;
+
+    for (const command of cFormat) {
+      if (escaped) {
+        buf.push(command);
+        escaped = false;
+        continue;
+      }
+      if (i >= ii) {
+        break;
+      }
+      switch (command) {
+        case "?":
+          buf.push(currCase(cSource.charAt(i++)));
+          break;
+        case "X":
+          while (i < ii) {
+            const code = cSource.charCodeAt(i++);
+            if (
+              (LA <= code && code <= LZ) ||
```

Comment:
Wouldn't it be possible to use `chatAt` instead so this can be simplified to `code >= 'a' && code <= 'z'`? Then you can remove the `limits` variable above and it seems more readable to me to remove the charcode conversion if it's not necessary.